docusaurus/blog/2023-02-15-odra-cosmos.md

```diff
@@ -0,0 +1,260 @@
+    use super::*;
+
+    #[test]
+    fn deploy() {
+        let counter = CounterDeployer::init(10);
+        assert_eq!(10, counter.get_value());
+    }
+
+    #[test]
+    fn increment() {
+        let mut counter = CounterDeployer::init(10);
+        counter.increment();
+        assert_eq!(11, counter.get_value());
+    }
+}
+```
+But wait, I mentioned CosmWasm, did I?
+
+Here the beauty of Odra comes into play.
+
+Let's use `cargo-odra`.
+```bash
+cargo odra build -b cosmos
+```
+And... that's it, congratulations! We have just written and build our first CosmWasm contract.
+As you see, it is nothing different from building a contract for Casper. No additional code, we only changed the `-b` flag.
+
+## Deploy
+We have just built a wasm file, but is it really a fully functional contract?
+
+As a battlefield let's choose [Juno Network] (if you would like to read more about smart contract development on Juno read this [Quick Start tutorial]). This is an arbitrary choice, each client is built upon a so-called Wasm Zone [wasmd], and its interface is alike.
+
+Assuming you already know how to interact with Juno testnet, let's move to the fun part.
+
+But before we go, to keep things simple, let's prepare a [justfile]. It'll make our interactions with the blockchain much  easier. See [full version].
+
+```justfile title=justfile
+NODE := "--node https://rpc.uni.juno.deuslabs.fi:443"
+CHAIN_ID := "--chain-id uni-6"
+QUERY_FLAGS := NODE + " " + CHAIN_ID
+TRANSACTION_DEFAULTS := "--gas-prices 0.025ujunox --gas auto --gas-adjustment 1.3 --broadcast-mode block"
+EXEC_FLAGS := NODE + " " + CHAIN_ID + " " + TRANSACTION_DEFAULTS
+
+get-address NAME:
+    junod keys show {{NAME}} | grep -o juno.*
+
+store-wasm WASM_PATH SENDER:
+    junod tx wasm store \
+    {{WASM_PATH}} --from {{SENDER}} {{EXEC_FLAGS}}
+
+init-contract CODE_ID VALUE SENDER CONTRACT_NAME:
+    junod tx wasm instantiate \
+    {{CODE_ID}} \
+    `just run-args-parser '{"name": "init", "args": [ { "value" : {{VALUE}} }]}'` \
+    --label '{{CONTRACT_NAME}}' --from {{SENDER}} \
+    --admin `just get-address {{SENDER}}` \
+    {{EXEC_FLAGS}}
+
+exec-increment ADDRESS SENDER:
+    junod tx wasm execute \
+    {{ADDRESS}} \
+    `just run-args-parser '{"name": "increment"}'` \
+    --from {{SENDER}}  \
+    {{EXEC_FLAGS}}
+
+query-get-value ADDRESS:
+    junod q wasm contract-state smart {{ADDRESS}} \
+    `just run-args-parser '{"name": "get_value"}'` {{QUERY_FLAGS}}
+```
+
+Ok, we are ready to go.
+
+First, a CosmWasm contract needs to be stored, technically is not a contract yet. Like a larva waiting to morph into a butterfly (sorry for that).
+
+There are three ways to interact with a contract.
+1. Instantiate - in other words, a constructor call. Once the contract is instantiated, it gets an address.
+2. Execute - call an entrypoint that modifies the state.
+3. Query - read the contract's state.
+   
+Now, let's take a look at how to do it using the tools we have just prepared.
+
```

Comment:
Here I think it would be better to have each command in separate box and show some of results. It doesn't feel now that those are console.